File: lab6/lab6.pl
#! /usr/bin/perl
use strict;
use warnings;
use POSIX;
use List::Util qw(max min);

sub putoption {
   my $strike = shift;
   return sub {my $final = shift; return max(0, $strike - $final);};
}
sub calloption {
   my $strike = shift;
   return sub {my $final = shift; return max(0, $final - $strike);};
}

my %choosehash;
sub nchoosek {
   my $n = shift;
   my $k = shift;
   $k > $n and return 0;
   if ($k > floor($n / 2)) {
      $k = $n - $k;
   }
   my $key = "n" . $n . "k" . $k;
   if (exists $choosehash{$key}) {
      return $choosehash{$key};
   }
   my $sum = 1;
   for (1..$k) {
      $sum *= $n--;
      $sum /= $_;
   }
   $choosehash{$key} = $sum;
   return $sum;
}

sub optionValue {
   my ($periods, $s0, $u, $d, $r, $option) = @_;

   my $p = ($r - $d ) / ($u - $d);
   my $q = 1 - $p;

   my $sum = 0;
   for (0..$periods) {
      my $left = ($p ** $_) * ($q ** ($periods - $_));
      my $value = $option->($s0 * ($u ** $_) * ($d ** ($periods - $_)));
      $sum += $left * nchoosek($periods, $_) * $value;
   }

   return ((1 / $r) ** $periods) * $sum;
}

my $runs = 1020;
my $option = putoption(60);
for (1..980) {
   optionValue($runs, 50, 1.004, 1 / 1.004, 1.0001, $option);
}
print optionValue($runs, 50, 1.004, 1 / 1.004, 1.0001, calloption(60)) . "\n";
print optionValue(2, 75, 6/5, 4/5, 11/10, putoption(95)) . "\n";

File: lab6/lab6.fs
#light

let putoption strike final = max 0.0 (strike - final)
let calloption strike final = max 0.0 (final - strike)

let nchoosek (n : int) k = 
   if n < k then 0
   else
      let newK = if k > (n / 2) then n - k else k
      let myS = seq { for i in 1..newK -> (n-(i-1), i) } 
      Seq.fold (fun acc x -> acc * fst x / snd x ) 1 myS

let optionValue (periods : int) (s0 : double) 
                (u : double) (d : double) (r : double) 
                (opt : double -> double) =
   let p = (r - d) / (u - d)
   let q = 1.0 - p

   let myS = seq { 0..periods }
   let left iter = (p ** iter) * (q ** ((double periods) - iter))
   let right iter = (opt (s0 * (u ** iter) * (d ** ((double periods) - iter))))
   let sum = Seq.fold 
              (fun acc x -> 
               (left (double x)) * 
               (double (nchoosek periods x)) * 
               (right (double x)) + 
               acc)
              0.0 myS
   
   ((1.0 / r) ** (double periods)) * sum

optionValue 2 75.0 (6.0 / 5.0) (4.0 / 5.0) (11.0 / 10.0) (putoption 95.0)
   |> printfn "%f"
